File: core/src/main/java/io/github/shaksternano/borgar/core/media/reader/NoAudioReader.java
package io.github.shaksternano.borgar.core.media.reader;

import io.github.shaksternano.borgar.core.collect.ClosableIterator;
import io.github.shaksternano.borgar.core.collect.ClosableSpliterator;
import io.github.shaksternano.borgar.core.media.AudioFrame;
import io.github.shaksternano.borgar.core.media.MediaReaderFactory;

import java.io.File;
import java.io.InputStream;
import java.util.Collections;
import java.util.Spliterators;
import java.util.function.Consumer;

public class NoAudioReader extends BaseMediaReader<AudioFrame> {

    public static final NoAudioReader INSTANCE = new NoAudioReader();

    private NoAudioReader() {
        super("N/A");
    }

    @Override
    public AudioFrame readFrame(long timestamp) {
        throw new UnsupportedOperationException("No audio available");
    }

    @Override
    public AudioFrame first() {
        throw new UnsupportedOperationException("No audio available");
    }

    @Override
    public MediaReader<AudioFrame> reversed() {
        return this;
    }

    @Override
    public ClosableIterator<AudioFrame> iterator() {
        return ClosableIterator.wrap(Collections.emptyIterator());
    }

    @Override
    public void forEach(Consumer<? super AudioFrame> action) {
    }

    @Override
    public ClosableSpliterator<AudioFrame> spliterator() {
        return ClosableSpliterator.wrap(Spliterators.emptySpliterator());
    }

    @Override
    public void close() {
    }

    public enum Factory implements MediaReaderFactory<AudioFrame> {

        INSTANCE;

        @Override
        public MediaReader<AudioFrame> createReader(File media, String format) {
            return NoAudioReader.INSTANCE;
        }

        @Override
        public MediaReader<AudioFrame> createReader(InputStream media, String format) {
            return NoAudioReader.INSTANCE;
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/reader/NoAudioReader.kt
package io.github.shaksternano.borgar.core.media.reader

import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.media.AudioFrame
import io.github.shaksternano.borgar.core.media.AudioReaderFactory
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emptyFlow
import kotlin.time.Duration

object NoAudioReader : BaseAudioReader() {

    override val frameCount: Int = 0
    override val frameRate: Double = 0.0
    override val duration: Duration = Duration.ZERO
    override val frameDuration: Duration = Duration.ZERO
    override val audioChannels: Int = 0
    override val audioSampleRate: Int = 0
    override val audioBitrate: Int = 0
    override val loopCount: Int = 0

    override suspend fun readFrame(timestamp: Duration): AudioFrame =
        throw UnsupportedOperationException("No audio available")

    override fun asFlow(): Flow<AudioFrame> = emptyFlow()

    override suspend fun reversed(): MediaReader<AudioFrame> = this

    override suspend fun createChangedSpeed(speedMultiplier: Double): MediaReader<AudioFrame> = this

    override suspend fun close() = Unit

    class Factory(
        override val supportedFormats: Set<String>
    ) : AudioReaderFactory {
        override suspend fun create(input: DataSource): AudioReader = NoAudioReader
    }
}
